### ISSUE TITLE
Python 3.12

### ISSUE THREAD

**Issue by vermut** [piwheels/piwheels#346]
Could you please start supporting Python 3.12 too? We miss it on fresh HomeAssistant install very much.

https://github.com/pytries/marisa-trie/pull/102

**Comment by bennuttall** [comment]
No. As stated on the homepage and FAQ, we can only support Python versions provided by the OS.

When Debian Trixie comes out next year, it'll have 3.12 or 3.13 and we'll build for that.

**Comment by vermut** [comment]
Tough luck. Seems like HomeAssistant jumped the gun with 3.12.

**Comment by waveform80** [comment]
Just to add some detail to Ben's reply, the HomeAssistant OS images aren't based on RaspiOS; they're [buildroot based images](https://github.com/home-assistant/operating-system):

> Home Assistant Operating System is not based on a regular Linux distribution like Ubuntu. It is built using [Buildroot](https://buildroot.org/) and it is optimized to run Home Assistant

piwheels is a distro-specific to RaspiOS. Even if your OS just happens to have the same version of Python as RaspiOS, that's no guarantee anything from piwheels will work because versions of build dependency libraries may be entirely different. Simply put, if it isn't RaspiOS, piwheels doesn't support it.

**Comment by vermut** [comment]
Technically we talking about HomeAssistant Core that tries to install above-mentioned `marisa-trie`, and it's derived from `alpine3.19` in a long chain of `FROM`s. Which doesn't help me even more because I'm clearly in a wrong place as you are building Debians. Still wondering how it worked before.

Thank you guys for explanations.

**Comment by waveform80** [comment]
Ah, no we don't build debs -- we build wheels for pip. However, many of these wheels incorporate compiled code and we deliberately compile against the *system* libraries (from debs), not other things pulled from PyPI.

This results in wheels that (at least potentially) depend on the content of debs installed on the system. This is why on some of the piwheels project pages you find instructions that include `apt install` alongside `pip install`, like [numpy](https://www.piwheels.org/project/numpy/) where piwheels has looked at the wheel output and determined the apt packages the wheel depends on (fortran and blas).

While this isn't a particularly modern approach (everyone seems to think static linking is the answer to everything these days), it does result in minimal wheels that re-use system libraries, keeping memory consumption lower (important on 512MB boards like the Pi Zero where piwheels arguably has the greatest impact).

There are "pure Python" wheels on piwheels which have no apt dependencies, and arguably these are likely to be compatible with things other than RaspiOS. However, we can't practically support that use-case, so we're fairly brutal in declaring piwheels a strictly RaspiOS-related service and closing anything that doesn't match that criteria.
